### PR TITLE
Allow binary postdata in HTTPFile::write

### PR DESCRIPTION
**Pull request by ghost**
I'm not quite sure who uses this piece of hardcoded stuff (application/json)
Changed like this it's more general and not restricted nly for strings.

The place where it is called (and assumed to be JSON) has to be changed (add mime type)
Hopefully somebody can remember.... :-)

### CODE REVIEW

File: xbmc/filesystem/CurlFile.cpp [hunk 1]
```diff
@@ -540,7 +540,7 @@ void CCurlFile::SetCommonOptions(CReadState* state)
   if (m_postdataset)
   {
     g_curlInterface.easy_setopt(h, CURLOPT_POST, 1 );
-    g_curlInterface.easy_setopt(h, CURLOPT_POSTFIELDSIZE, m_postdata.length());
+    g_curlInterface.easy_setopt(h, CURLOPT_POSTFIELDSIZE, m_postdata.size());
```
Comment (FernetMenta) [inline xbmc/filesystem/CurlFile.cpp]: afaik there is no difference between size and length. why did you change it?

Comment (mapfau) [inline xbmc/filesystem/CurlFile.cpp]: oh, it is - /0 is end of counting if length is used

Comment (FernetMenta) [inline xbmc/filesystem/CurlFile.cpp]: hmm, all sources I found say that size and length are identical: http://en.cppreference.com/w/cpp/string/basic_string

Comment (mapfau) [inline xbmc/filesystem/CurlFile.cpp]: I could provide some binary data where length() is completely different to size().
But - I realy know this from experiences at work with UTF8.

Need the examle? Or trust me?

Comment (fritsch) [inline xbmc/filesystem/CurlFile.cpp]: Cited from cplusplus.com
> Both string::size and string::length are synonyms and return the exact same value.

I am curious :-) Could you add a CLog that really shows a difference between size and length?

Comment (mapfau) [inline xbmc/filesystem/CurlFile.cpp]: Maybe not a strictly formal answer, but I've seen the convention that 'size' refers to the number of bytes, and 'length' refers to the number of elements. So for a char array, the size and length are the same, but for arrays of other data types the size would be greater.

Comment (mapfau) [inline xbmc/filesystem/CurlFile.cpp]: will present an example this evening - I have to reactivate smoothstreammedia addon...

Comment (fritsch) [inline xbmc/filesystem/CurlFile.cpp]: >Note that string objects handle bytes without knowledge of the encoding that may eventually be used to encode the characters it contains. Therefore, the value returned may not correspond to the actual number of encoded characters in sequences of multi-byte or variable-length characters (such as UTF-8).

See - they do it exactly the same. Can you post a string where this is different? I am really only curious - as I did not see something like that in an std::string until now.

Comment (fritsch) [inline xbmc/filesystem/CurlFile.cpp]: Something else: https://curl.haxx.se/libcurl/c/CURLOPT_POSTFIELDSIZE.html we need the size in bytes anyways in the described usecase.

Comment (mapfau) [inline xbmc/filesystem/CurlFile.cpp]: Yes, this was also changed in this PR

Comment (fritsch) [inline xbmc/filesystem/CurlFile.cpp]: Yeah :-) We are not saying that your change is wrong but from the implementation of std::string it should not matter.

**Comment by mapfau** [comment]
Are there any objections against this? This one is necessary for smoothstreamingmedia inputstream development and It would make my life easier it is in the nightlies. currently I have to tell developers on plugins based on smoothstream that they have to merge / pull something. Not everytime a nice job :-)

**Comment by FernetMenta** [comment]
see discussion here: https://github.com/xbmc/xbmc/pull/1650
they mentioned already that it should have a more general solution. But as always, once quick hacks are merges, nobody cares anymore.

**Comment by Montellese** [comment]
This has been added a long time ago, see #1650. Looks like it might be used by some PVR addon(s)?

**Comment by Montellese** [comment]
Sorry, wrong button.

**Comment by mapfau** [comment]
then we should remove CHTTPFile completely and give the addon interface the possibility to post (as you (@fernetmenta) mentioned already some month ago??

**Comment by FernetMenta** [comment]
makes sense to me

**Comment by arnova** [comment]
That was also my suggestion back then when this HTTPFile hack was introduced.

**Comment by mapfau** [comment]
Then somebody has to remeber at which place it is used. The word PVR was named in the ticked. Is it maybe obsolete that PVR uses POST requests?

**Comment by mapfau** [comment]
continued @ https://github.com/xbmc/xbmc/pull/9621
